src/Microsoft.Azure.ServiceBus/Core/MessageReceiver.cs

```diff
@@ -960,5 +1009,48 @@ void ThrowIfNotPeekLockMode()
 
             MessagingEventSource.Log.RegisterOnMessageHandlerStop(this.ClientId);
         }
+
+        public void RegisterPlugin(ServiceBusPlugin serviceBusPlugin)
+        {
+            if (serviceBusPlugin == null)
+            {
+                throw new ArgumentNullException(nameof(serviceBusPlugin), Resources.ArgumentNullOrWhiteSpace.FormatForUser(nameof(serviceBusPlugin)));
+            }
+            if (this.RegisteredPlugins == null)
+            {
+                this.RegisteredPlugins = new List<ServiceBusPlugin>();
+            }
+            else if (this.RegisteredPlugins.Any(p => p.GetType() == serviceBusPlugin.GetType()))
+            {
+                throw new ArgumentException(nameof(serviceBusPlugin), Resources.PluginAlreadyRegistered.FormatForUser(nameof(serviceBusPlugin)));
+            }
+            this.RegisteredPlugins.Add(serviceBusPlugin);
+        }
+
+        public void UnregisterPlugin(ServiceBusPlugin serviceBusPlugin)
```

Comment:
Yeah, I cant find my own comment. Reposting it here:

>I partly agree with @Sean here. I think we should be able to register plugin only during creation of Queue unless we see a strong reason to change things dynamically. But at the same time, we might want to have UnregisterPlugin or ClosePlugin equivalent so that plugin can probably release some resources of its own if required. But this would require us to expose additional methods on plugin. So in effect, we might end up having two more functions OnPluginRegistered and OnPluginUnregistered methods in ServiceBusPlugin

>But now to think of it, i can think of scenarios where we may want to unregister plugin. Lets say we have a plugin to tackle inter-op scenarios for users to switch from old client to new client. At some point in time they realise that they are done with the interop handling (upgrade scenarios), and now want to unregister the plugin.